### ISSUE TITLE
Codelens flashing after folding

### ISSUE THREAD

**Issue by rebornix** [microsoft/vscode#121516]
Issue Type: <b>Bug</b>

* Create an untitled file
* Change language mode to TS
* Paste content of `vscode.proposed.d.ts` into it
* Run Fold All

VS Code version: Code - Insiders 1.56.0-insider (Universal) (ce791d27dc9889a392f4d29abefe9c53eb0b5129, 2021-04-16T11:23:31.851Z)
OS version: Darwin x64 20.3.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz (16 x 2400)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 2, 2|
|Memory (System)|32.00GB (0.68GB free)|
|Process Argv|--enable-proposed-api ms-vscode.vscode-notebook-history --enable-proposed-api vscode.notebook-editing --enable-proposed-api vscode.vscode-test-resolver --enable-proposed-api donjayamanne.kusto --crash-reporter-id 1a907bbd-fa02-4595-81e0-63e1648afb31|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by jrieken** [comment]
@aeschli Is there a way to "know" when folding has happened?

**Comment by aeschli** [comment]
Need @alexdima consulting here. Hidden line state and changes are well hidden in the editor. The folding editor contribution uses `editor.setHiddenAreas` to tell the editor what is hidden and I'm not aware of events from the editor to the outside.

**Comment by alexdima** [comment]
The editor core is not aware of folding, but has API that folding uses, to hide lines, called `setHiddenAreas`.

The hidden areas can be observed when calling `editor.getVisibleRanges()`. The returned ranges will have gaps for the lines which are hidden. But that method is only for the viewport. @jrieken Would you need a global `getHiddenAreas()`?

**Comment by jrieken** [comment]
> Is it a problem in the core? 

Yeah, that's my assumption. There is no notion of folding or hidden areas in the code lens code and I would assume that the core hides widgets in hidden regions and I don't have to do anything, right?

**Comment by jrieken** [comment]
The only thing I maybe should know is when folding has occurred. This issue is surely related to me updating a widget that's inside a region (guess, not debugging). This is my theory:

* spaces for lens is reserved (that's quick cos from cache)
* request to resolve is send
* folding occurs
* resolve is done and updates all lenses

So, I could cancel the last step (or parts of that) because folding occurred but I still thing there is something wrong in core
